applications/ConvectionDiffusionApplication/python_scripts/convection_diffusion_solver.py

```diff
@@ -3,19 +3,28 @@
 # Importing the Kratos Library
 import KratosMultiphysics
 
+# Auxiliary function to check the parallel type at runtime
+#TODO: Delete this once we come up with the final factory-based design
+def _CheckIsDistributed():
+    if KratosMultiphysics.ParallelEnvironment.HasDataCommunicator("World"):
```

Comment:
Please regular people, avoid this comment

@roigcarlo Is this a Jojo reference?

Reply:
It always has been.

P.s: Oh... now I will see it every time I use MPI thx...

Reply:
This also doesn't work

If Kratos runs in MPI, then also this solver will always run in MPI, there is no way to choose

Reply:
Agree. But right now there is no technical solution to achieve what you pointed out. I think we can merge it as it is to then improve it in a future PR.

Reply:
Then the other only option would be to pass the comm via json no? I can't see any other way

Reply:
> Then the other only option would be to pass the comm via json no? I can't see any other way

The point of this is that it would imply to change the writing of the `ProjectParameters.json`.

Reply:
> Then the other only option would be to pass the comm via json no? I can't see any other way

#8613

;)

Reply:
You can merge if you need it now, I think we can improve this in the future

Reply:
> > Then the other only option would be to pass the comm via json no? I can't see any other way
> 
> The point of this is that it would imply to change the writing of the `ProjectParameters.json`.

https://github.com/KratosMultiphysics/Kratos/blob/40a78b5864f1dc463030db28914be354bd153408/kratos/mpi/python_scripts/distributed_import_model_part_utility.py#L13-L16

you can specify it (or not, the default is `world`

Reply:
Maybe I'm wrong, but this would require the user or the GUI to set the communicator name in the `model_import_settings`. On top of that, this is not know when doing the import as the json settings hasn't been parsed yet. Am I wrong?

Reply:
But in this example, you know you are in a parallel context. We are back to having a dedicated file ( in this case is the import_model_part_utility vs distributed_import_model_part_utility)

This is good for, for example lines 269, 290, etc.. which are like:

```Python
# If MPI distributed, add the PARTITION_INDEX
if _CheckIsDistributed():
    target_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.PARTITION_INDEX)
```

Having to parse the json file in order to know if we are in a parallel context is too excessive, and we don't really care if we are going to end up executing the solver in parallel or not, its just that if we do (independently on the communicator used) we will need to have those modules imported. Or at least is how I see it. Again I agree that this is not the best solution, but asking for the default comm, or any comm for what matters through the default in this particular place should not be a problem.

Reply:
How would you get around specifying whether you are in MPI or not in the json?

I don’t think that this would be a good idea, IMO the parallel stuff is too complex to hide completely 

and setting the data comm to sth else than world is a very special yet necessary feature.
Otherwise as soon as you have a solver that doesn’t support MPI you cannot run other solvers in parallel
My default example: CoSimulation with MPI fluid and serial structure

Reply:
I am not sure to follow you now (i think we are discussing different things)

I assume the problematic scenarios are the following:

1 - **You run with mpi, you have a solver with mpi support:**
In this case. First, you need to know first if the mpi imports are nedded or not. If your solver has mpi compatibility you will put the code there, otherwise you won't, and you will import that independetly of the fact that you run with mpi or not in the end. Second, you need to know of you want to run with mpi or not, which can be set in the json by setting "serial", "world", or watherver communicator you are interested in.

Hance the code of the solver would look like this:
```Python
import Kratos
if Kratos.GetDefaultComm().IsDistributed()  # This could be replaced by Kratos.ParallelEnv.CanWeRunInPrallel() witout the need of knowing if we are going to do it, just if there is the possibility
    we_are_in_parallel = True
    import Kratos.mpi_related_things

# Now we need to know if our case runs in parallel or not
# Here the suer selects which comm he wants, "serial", "world", "custom...". If it does not exists we default to whatever we decide. In this snipet I default to the fastest framework possible. (serial if not mpirun was used, world if we are un parallel and the solver suports it)
if settings["model_import_settings"].Has("data_communicator_name"): 
     data_comm_name = settings["model_import_settings"]["data_communicator_name"].GetString() 
 else: 
     if  we_are_in_parallel 
         data_comm_name = "World" 
     else 
         data_comm_name = "Serial" 

communicator = Kratos.GetCommunicator(data_comm_name )

# From this point onwards, we have the correct communicator, either being the serial one, the world one, or the one the user chooses.
```

2 - **You run with mpi, you have a solver without mpi support**:
Here you don't even check if you want to use mpi, you cant, so the parse of the json file is not done, you don't need to query for the first `Kratos.GetDefaultComm().IsDistributed()` because you don't have extra imports.

3 - **You run without mpi, you have a solver with mpi support**:
Here you won't be able to retrive  any default communicator that return IsDistributed() = True, so you will never set the `we_are_in_parallel = True`, and this will force `data_comm_name = "Serial" `

I honestly believe this covers all the cases, and again I agree that refactoring this such that this problem never happens in the first place would be better, but honestly I don't see how this implementation breaks anything.

Reply:
My issue with the current design is that when Kratos runs in MPI you force that the solver also uses MPI, one cannot select the serial execution

Your draft is already better but what I strongly dislike is the unconditional imports of MPI things. We put a lot of effort in the past to only import when some apps are really needed.

Think that you have solver A and solver B, both support MPI: Solver A (e.g. DEM) runs in MPI but doesn't require "nasty" apps like Trilinos. Solver B however runs in serial, but if it would run in parallel then it would use e.g. Trilinos.
With your design you force that the user compiles the TrilinosApp even though it is not used

Thats why I suggested the backed idea, where you only import things you actually need

Reply:
Give me time until the end of the week to think of a better way of doing this :)

Reply:
I thought abt it and the only way I see the MPI related imports can be done is not at the top of the file, as we don't know yet if the solver will run in mpi or not (emphasizing again that Kratos running in MPI does NOT mean that every solver runs in MPI)

Hence the fct _CheckIsDistributed should also be used to import the MPI dependencies (e.g. TrilinosApp)
I agree that this is a bit cumbersome but it should become much simpler with the factories